src/helpers/messaging.rs

```diff
@@ -85,15 +88,68 @@ pub struct Gateway {
     tx: mpsc::Sender<ReceiveRequest>,
     envelope_tx: mpsc::Sender<SendRequest>,
     control_handle: JoinHandle<()>,
+    shutdown_tx: oneshot::Sender<()>,
+    channels: Arc<Mutex<HashMap<Step, NonZeroUsize>>>,
 }
 
 pub(super) type SendRequest = (ChannelId, MessageEnvelope);
 
+#[derive(Clone, Copy, Debug)]
+pub enum TotalRecords {
+    Unspecified,
+    Specified(NonZeroUsize),
+
+    /// Total number of records is not well-determined. When the record ID is
+    /// counting solved_bits attempts. The total record count for solved_bits
+    /// depends on the number of failures.
+    ///
+    /// The purpose of this is to waive the warning that there is a known
+    /// number of records when creating a channel. If the warning is firing
+    /// and the total number of records is knowable, prefer to specify it
+    /// rather than use this to waive the warning.
+    Indeterminate,
+}
+
+impl TotalRecords {
+    #[must_use]
+    pub fn is_unspecified(&self) -> bool {
+        matches!(self, &TotalRecords::Unspecified)
+    }
+
+    #[must_use]
+    pub fn is_indeterminate(&self) -> bool {
+        matches!(self, &TotalRecords::Indeterminate)
+    }
+}
+
+impl From<usize> for TotalRecords {
+    fn from(value: usize) -> Self {
+        match NonZeroUsize::new(value) {
+            Some(v) => TotalRecords::Specified(v),
+            None => TotalRecords::Unspecified,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct TotalRecordsUnknownError;
+
+impl TryInto<NonZeroUsize> for TotalRecords {
+    type Error = TotalRecordsUnknownError;
+    fn try_into(self) -> Result<NonZeroUsize, Self::Error> {
+        match self {
+            TotalRecords::Specified(value) => Ok(value),
+            _ => Err(TotalRecordsUnknownError),
```

Comment:
a nit: it does not really unknown to us at this point. We could say exactly why we can't convert